code/modules/vote/vote_datum.dm

```diff
@@ -0,0 +1,198 @@
+	var/list/voted = list()
+
+
+/datum/vote/New(_initiator, _question, list/_choices, _is_custom = FALSE)
+	if(SSvote.active_vote)
+		CRASH("Attempted to start another vote with one already in progress!")
+
+	if(_initiator)
+		initiator = _initiator
+	if(_question)
+		question = _question
+	if(_choices)
+		choices = _choices
+
+	is_custom = _is_custom
+
+	// If we have no choices, dynamically generate them
+	if(!length(choices))
+		generate_choices()
+
+/datum/vote/proc/start()
+	var/text = "[capitalize(vote_type_text)] vote started by [initiator]."
+	if(is_custom)
+		vote_type_text = "custom"
+		text += "\n[question]"
+		if(usr)
+			log_admin("[capitalize(vote_type_text)] ([question]) vote started by [key_name(usr)].")
+
+	else if(usr)
+		log_admin("[capitalize(vote_type_text)] vote started by [key_name(usr)].")
+
+	log_vote(text)
+	started_time = world.time
+	announce(text)
+
+/datum/vote/proc/remaining()
+	return max(((started_time + GLOB.configuration.vote.vote_time) - world.time), 0)
+
+
+// Returns the result
+/datum/vote/proc/calculate_result()
+	switch(vote_result_type)
+		if(VOTE_RESULT_TYPE_MAJORITY)
+			if(!length(voted))
+				to_chat(world, "<span class='interface'>No votes were cast. Do you all hate democracy?!</span>") // shame them
+				return null
+
+			var/list/results = list()
+
+			// Count up all votes
+			for(var/ck in voted)
+				if(voted[ck] in results)
+					results[voted[ck]]++
+				else
+					results[voted[ck]] = 1
+
+			// Get the biggest vote count, since we can also use this to pick tiebreaks
+			var/maxvotes = 0
+			for(var/res in results)
+				maxvotes = max(results[res], maxvotes)
+
+			var/list/winning_options = list()
+
+			for(var/res in results)
+				if(results[res] == maxvotes)
+					winning_options |= res
+
+			// Print all results
+			for(var/res in results)
+				if(res in winning_options)
+					// Make it stand out
+					to_chat(world, "<span class='info'><code>[res]</code> - [results[res]] vote\s</span>")
+				else
+					// Make it normal
+					to_chat(world, "<span class='interface'><code>[res]</code> - [results[res]] vote\s</span>")
+
+				// And log it to the DB
+				if(!is_custom)
+					SSblackbox.record_feedback("nested tally", "votes", results[res], list(vote_type_text, res), ignore_seal = TRUE)
+
+			if(length(winning_options) > 1)
+				var/random_dictator = pick(winning_options)
+				to_chat(world, "<span class='interface'><b>Its a tie between [english_list(winning_options)]. Picking <code>[random_dictator]</code> at random.</b></span>") // shame them
+				return random_dictator
+
+			// If we got here there must only be one thing in the list
+			var/res = winning_options[1]
+
+			if(res in choices)
+				to_chat(world, "<span class='interface'><b><code>[res]</code> won the vote.</b></span>")
+				return res
+
+			to_chat(world, "<span class='interface'>The winner of the vote ([sanitize(res)]) isnt a valid choice? What the heck?</span>")
```

Comment:
might be good to print a runtime? not that this should actually happen.

Reply:
Could do